⚠ Status condition clean up

anik120 opened the pull request:
[RFC: ClusterCatalog Status Conditions](https://docs.google.com/document/d/1Kg8ovX8d25OqGa5utwcKbX3nN3obBl2KIYJHYz6IlKU/edit) implementation.

Closes [363](https://github.com/operator-framework/catalogd/issues/363) Closes [364](https://github.com/operator-framework/catalogd/issues/364) Closes [365](https://github.com/operator-framework/catalogd/issues/365)

<!--
Please prefix the title of this PR with one of the following icons:

    * ⚠ (:warning:, major/breaking change)
    * ✨ (:sparkles:, minor/compatible change)
    * 🐛 (:bug:, patch/bug fix)
    * 📖 (:book:, docs)
    * 🌱 (:seedling:, other)

-->

everettraven commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -258,6 +238,9 @@ func (r *ClusterCatalogReconciler) needsUnpacking(catalog *v1alpha1.ClusterCatal
 	if !r.Storage.ContentExists(catalog.Name) {
 		return true
 	}
+	if !servingStatusConditionExists(catalog) {
+		return true
+	}
```
Do we actually need this? Don't we already read from our source of truth by checking if the content exists?

everettraven replied:
Looking at the tests I see what the intention was here. I think like @joelanford called out in the sync we had yesterday, we should probably have an earlier check in `reconcile` for "IF content exists and is being served, set serving to true"

I don't think the content existing but the serving status missing means we should unpack.

anik120 replied:
Right, I looked into this a bit more after the sync, and here's the conclusion I deviated towards: 

`r.needsUnpacking` is the earliest call in the `reconcile` function (the if conditions before that are either setting the finalizer for the first time the CR is created, or cleaning up stuff when the CR is deleted, and both are exiting after doing their job). r.needsUnpacking calls `r.StorageExists`, and if false is returned, `unpackNeeded = true` is signaled. 

So adding this check here, is essentially doing this: 

> "IF content exists and is being served, set serving to true"

Looking at the code: 

```
if catalog.Status.ResolvedSource == nil {
		return true
	}
	if !r.Storage.ContentExists(catalog.Name) {
		return true
	}
	if !servingStatusConditionExists(catalog) {
		return true
	}
```
* If the content doesn't exist, then we need to run through the reconcile loop, unpack, store and set `Serving=true`  
* If the content exists, it means the content was unpacked in a previous loop, and that loop should have set the condition `Serving=true` 
* If for some reason the condition `Serving=true` was not set, then we're signaling, "go ahead and continue with the `reconcile` loop anyway". In this case, `souce.Unpack` is called again, but it doesn't actually store because the [image_registry_client source sees that the content path already exists](https://github.com/operator-framework/catalogd/blob/main/internal/source/image_registry_client.go#L98), so the loop continues, and the `Serving=True` condition is eventually set this time (if not, then of course it'll be retried in the next loop and so on..).

So it achieves 

> "IF content exists and is being served, set serving to true"

without having to worry about dealing with calling `meta.SetStatusCondition` on the `Serving` condition twice.

everettraven replied:
While I understand the thinking here, I still don't think we should be triggering unpack in this case. Unpacking again means we are making at least another request to the image registry to check if it is the same digest and then we are going through the storage process again for the same information. This is unnecessary cycles, plus it is safe to call `meta.SetStatusCondition` as many times as we want on the same condition because it guards changes to the last transition time on a change in the status (i.e True, False, Unknown): https://github.com/kubernetes/apimachinery/blob/2465dc5239ab8827a637148a78b380c278b4a5f4/pkg/api/meta/conditions.go#L44-L52

joelanford replied:
> as many times as we want on the same condition because it guards changes to the last transition time on a change in the status (i.e True, False, Unknown)

This is true as long as we never do something like this:
1. the condition is True when we read it from the cache at the beginning of `Reconcile`
2. we check storage and see that nothing is there, so we set it to `False`.
3. because nothing is in storage, we have to unpack. that happens successfully, so we set it back to `True`

If we do that, the lastTransitionTime in the condition will be updated even though an outside observer will see no change in the status. It will be `True` at all times. The `False` status was only in memory and never persisted to etcd.

joelanford replied:
> I still don't think we should be triggering unpack in this case.

Definitely agree with Bryce here. A condition being stale does not mean our data is stale. I suspect we do not need to read our status _at all_. Anywhere we are reading status should be highly scrutinized and likely refactored/removed.

everettraven commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -180,73 +176,57 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
 }
 
-func updateStatusUnpackPending(status *v1alpha1.ClusterCatalogStatus, result *source.Result) {
+func updateStatusProgressing(status *v1alpha1.ClusterCatalogStatus) {
```
(Optional) Nit: instead of having multiple functions for setting the progressing status we could probably get away with a single one that sets `Progressing` to `True` if the error provided is `nil`

everettraven commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -180,73 +176,57 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
 
-func updateStatusUnpackFailing(status *v1alpha1.ClusterCatalogStatus, err error) error {
-	status.ResolvedSource = nil
+func updateStatusNotServing(status *v1alpha1.ClusterCatalogStatus) {
```
If we are no longer serving, should we also unset the content url?

anik120 replied:
Good point

everettraven commented on internal/controllers/core/clustercatalog_controller_test.go:
```diff
@@ -535,7 +416,9 @@ func TestCatalogdControllerReconcile(t *testing.T) {
-			diff := cmp.Diff(tt.expectedCatalog, tt.catalog, cmpopts.IgnoreFields(metav1.Condition{}, "Message", "LastTransitionTime"))
+			diff := cmp.Diff(tt.expectedCatalog, tt.catalog,
+				cmpopts.IgnoreFields(metav1.Condition{}, "Message", "LastTransitionTime"),
+				cmpopts.SortSlices(func(a, b metav1.Condition) bool { return a.Type < b.Type }))
```
is a deterministic ordering of the condition types necessary for comparison? I was under the impression the `cmp` package can handle unsorted comparisons.

anik120 replied:
Unfortunately looks like it is. I just discovered the `cmpopts.SortSlices` function today because the slice ordering was causing the test to fail.

everettraven commented on internal/controllers/core/clustercatalog_controller_test.go:
Add a test that marks `Progressing` as `False` with reason `Unrecoverable` when the mock unpacker returns an `errors.Unrecoverable`?

grokspawn replied:
+1

anik120 replied:
I was thinking about this to solve the codecov complaint of a -2.42% coverage diff for the patch, but then got lazy. Now that you mention it.....adding it 😁

anik120 replied:
Added it, fyi this test revealed that we were not actually setting the `Reconciler` err to nil when we receive an unrecoverable error, because we were wrapping the `Unrecoverable` error inside a errors.Error [here](https://github.com/operator-framework/catalogd/blob/main/internal/controllers/core/clustercatalog_controller.go#L142) :) I changed that to just return the error we get from `Unpack`, that way the `error.As(*unrecoverable)` returns true

grokspawn commented:
Seems like we should also update [here](https://github.com/operator-framework/catalogd/blob/6584e7e3a1ff5688b375505f1c248e68ea6604f7/hack/scripts/demo-script.sh#L22) and [here](https://github.com/operator-framework/catalogd/blob/6584e7e3a1ff5688b375505f1c248e68ea6604f7/hack/scripts/gzip-demo-script.sh#L12) and update the demos.
If you prefer, we can do this as a follow-on issue, but we should generally strive to keep the demos up to date.

ankitathomas commented on api/core/v1alpha1/clustercatalog_types.go:
```diff
@@ -26,15 +26,17 @@ type SourceType string
+	ReasonSucceeded     = "Succeeded"
+	ReasonRetrying      = "Retrying"
+	ReasonUnrecoverable = "Unrecoverable"
 
```
Can you update the printcolumn  annotation from the RFC? 

```
//+kubebuilder:printcolumn:name="Serving",type=string,JSONPath=`.status.conditions[?(@.type=="Serving")].status`
```

anik120 replied:
Sure thing! Added PTAL!

grokspawn commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -139,24 +140,18 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
 	unpackResult, err := r.Unpacker.Unpack(ctx, catalog)
 	if err != nil {
-		return ctrl.Result{}, updateStatusUnpackFailing(&catalog.Status, fmt.Errorf("source bundle content: %v", err))
+		return ctrl.Result{}, updateStatusProgressError(&catalog.Status, fmt.Errorf("source bundle content: %v", err))
```
Doesn't an unpack failure also need to set `Serving := False/Unavailable`, i.e. `updateStatusNotServing`?

everettraven replied:
Not necessarily. If we've previously unpacked and are polling for an update we would still be serving the older contents

ankitathomas replied:
does that still hold if we don't delete what we're currently serving (i.e, during an upgrade)?

grokspawn replied:
Wouldn't we want to do something like
```go
if err != nil {
       if !r.storage.ContentExists(catalog.Name) {
           updateStatusNotServing(&catalog.Status)
        }
	return ctrl.Result{}, updateStatusProgressError(&catalog.Status, fmt.Errorf("source bundle content: %v", err))        
}
```
I couldn't find anything that convinced me that we start with Serving := False/Unavailable before the first unpack attempt.

everettraven replied:
If we want the `Serving` condition set on every reconcile it probably makes sense to do something like that. I think there are generally two schools of thought here (based on conversations I've had with @joelanford on status conditions):
- An unset condition should be interpreted as in an "Unknown" state. This should probably be interpreted the same as Serving := False. We could explicitly document the absence of this condition meaning it is not serving.
- Always set every condition on every reconcile.

I'm not super opinionated on an approach here and am fine with either one. In the current state of the codebase I think the first school of thought is exercised

grokspawn replied:
Yeah, happy to let this one go unmodified if we feel we're already giving a clear (implicit) indication.

ankitathomas reviewed (commented):
great work! 

There are a few more mentions of the Unpacking condition we should probably clean up in the README.md, docs/fetching-catalog-contents.md, test/tools/imageregistry/pre-upgrade-setup.sh and hack/scripts/gzip-demo-script.sh and hack/scripts/demo-script.sh. 

We'd want to at least update the pre-upgrade-setup script since that's part of upgrade testing, but it'd be good to keep everything else up to date

anik120 commented:
> great work!
> 
> There are a few more mentions of the Unpacking condition we should probably clean up in the README.md, docs/fetching-catalog-contents.md, test/tools/imageregistry/pre-upgrade-setup.sh and hack/scripts/gzip-demo-script.sh and hack/scripts/demo-script.sh.
> 
> We'd want to at least update the pre-upgrade-setup script since that's part of upgrade testing, but it'd be good to keep everything else up to date

Thanks for pointing this out @ankitathomas ! I updated most of it, except for the `pre-upgrade-setup.sh` file. Correct me if I'm wrong, but this has to remain unchanged in this PR, since `main` (pre this commit) still has the `Unpacking` condition, which mean the script will expect that condition. There needs to be a follow up PR, that has to change this script. 

Or did I get that wrong?  The pre-upgrade test is passing with this setting though....


Also, I'll look into the updating of the actual demo file after lunch, but for now I ran into this issue, not sure if @grokspawn can help

```
make demo-update           
hack/scripts/generate-asciidemo.sh
hack/scripts/generate-asciidemo.sh: line 40: [: /var/folders/9p/q0l_ddk11jqd05vb098xk9f40000gn/T/d.mVVVtAWPPU: binary operator expected
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  5481  100  5481    0     0  17382      0 --:--:-- --:--:-- --:--:-- 17400
curl: (6) Could not resolve host: generate-asciidemo.DTJEN
chmod: generate-asciidemo.DTJEN/asciinema-rec_script: Not a directory
/var/folders/9p/q0l_ddk11jqd05vb098xk9f40000gn/T/d.mVVVtAWPPU: line 155: generate-asciidemo.DTJEN/asciinema-rec_script: Not a directory
asciinema: /Users/anbhatta/go/src/github.com/operator-framework/catalogd/hack/scripts/demo-script.sh already exists, aborting
asciinema: use --overwrite option if you want to overwrite existing recording
asciinema: use --append option if you want to append to existing recording
usage: asciinema [-h] [--version] {rec,play,cat,upload,auth} ...
asciinema: error: unrecognized arguments: generate-asciidemo.DTJEN/catalogd-demo.cast
hack/scripts/generate-asciidemo.sh: line 17: [: /var/folders/9p/q0l_ddk11jqd05vb098xk9f40000gn/T/d.mVVVtAWPPU: binary operator expected
make: *** [demo-update] Error 2
```

ankitathomas commented:
Everything looks good!

> There needs to be a follow up PR, that has to change this script.

If we're ok with the test breaking before we're done with the API review work, it may be easier to have this change on your PR and cut a new release immediately after, but you're right - we can also update the upgrade-e2e script in a followup PR if we want to keep the test passing.

everettraven commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -139,24 +140,18 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
 	unpackResult, err := r.Unpacker.Unpack(ctx, catalog)
 	if err != nil {
-		return ctrl.Result{}, updateStatusUnpackFailing(&catalog.Status, fmt.Errorf("source bundle content: %v", err))
+		return ctrl.Result{}, updateStatusProgressing(catalog, err)
```
it should be okay to still wrap the error if we think it is necessary. I think we just need to update the `%v` directive to be `%w` for `errors.As()` to pickup the unrecoverable error - good catch here though!

joelanford replied:
Yes, please on adding helpful context everywhere. That is invaluable when trying to debug from logs.

everettraven commented on internal/controllers/core/clustercatalog_controller_test.go:
```diff
@@ -29,9 +30,16 @@ type MockSource struct {
 
 	// shouldError determines whether or not the MockSource should return an error when MockSource.Unpack is called
 	shouldError bool
+
+	// shouldErrorUnrecoverble determines whether or not the MockSource should return an error when MockSource.Unpack is called
+	// is Unrecoverable
+	shouldErrorUnrecoverable bool
```
Could update this to just have a field of the `error` type and set the exact error on each test that requires an error

everettraven commented on internal/controllers/core/clustercatalog_controller_test.go:
```diff
@@ -29,9 +30,16 @@ type MockSource struct {
 }
 
 func (ms *MockSource) Unpack(_ context.Context, _ *catalogdv1alpha1.ClusterCatalog) (*source.Result, error) {
+	if ms.shouldErrorUnrecoverable {
```
If we made the `MockSource` have a single field for the error I think we could simplify this to entire method to:
```go
func (ms *MockSource) Unpack(_ context.Context, _ *catalogdv1alpha1.ClusterCatalog)(*source.Result, error) {
        return ms.result, ms.err
}
```

grokspawn commented:
> Correct me if I'm wrong, but this has to remain unchanged in this PR

IMO, we're making these changes in full awareness that we are breaking the API.  We have opted not to create an interim API version to avoid it, and in the past we've committed and tried to cut a new release of the component to get the test passing again.

everettraven commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -180,73 +176,60 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
-		Message: result.Message,
-	})
+func updateStatusProgressing(catalog *v1alpha1.ClusterCatalog, err error) error {
+	var unrecov *catalogderrors.Unrecoverable
+	if err == nil {
+		catalog.Status.ResolvedSource = nil
+		meta.SetStatusCondition(&catalog.Status.Conditions, metav1.Condition{
+			Type:    v1alpha1.TypeProgressing,
+			Status:  metav1.ConditionFalse,
+			Reason:  v1alpha1.ReasonSucceeded,
+			Message: fmt.Sprintf("Successfully unpacked and stored content from %s", catalog.Spec.Source.Image.Ref),
+		})
+	} else if errors.As(err, &unrecov) {
+		meta.SetStatusCondition(&catalog.Status.Conditions, metav1.Condition{
+			Type:    v1alpha1.TypeProgressing,
+			Status:  metav1.ConditionFalse,
+			Reason:  v1alpha1.ReasonUnrecoverable,
+			Message: err.Error(),
+		})
+		return err
+	} else {
+		meta.SetStatusCondition(&catalog.Status.Conditions, metav1.Condition{
+			Type:    v1alpha1.TypeProgressing,
+			Status:  metav1.ConditionTrue,
+			Reason:  v1alpha1.ReasonRetrying,
+			Message: err.Error(),
+		})
+		return err
+	}
+	return nil
```
Missed this in my last review round, but would returning early make sense here instead of this longer if-elseif-else chain?
```suggestion
	var unrecov *catalogderrors.Unrecoverable
        if errors.As(err, &unrecov) {
		meta.SetStatusCondition(&catalog.Status.Conditions, metav1.Condition{
			Type:    v1alpha1.TypeProgressing,
			Status:  metav1.ConditionFalse,
			Reason:  v1alpha1.ReasonUnrecoverable,
			Message: err.Error(),
		})
		return err
	}
	
	if err != nil {
		meta.SetStatusCondition(&catalog.Status.Conditions, metav1.Condition{
			Type:    v1alpha1.TypeProgressing,
			Status:  metav1.ConditionTrue,
			Reason:  v1alpha1.ReasonRetrying,
			Message: err.Error(),
		})
		return err
	}
	
	catalog.Status.ResolvedSource = nil
	meta.SetStatusCondition(&catalog.Status.Conditions, metav1.Condition{
		Type:    v1alpha1.TypeProgressing,
		Status:  metav1.ConditionFalse,
		Reason:  v1alpha1.ReasonSucceeded,
		Message: fmt.Sprintf("Successfully unpacked and stored content from %s", catalog.Spec.Source.Image.Ref),
	})
	return nil
```

everettraven commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -180,73 +176,60 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
+func updateStatusProgressing(catalog *v1alpha1.ClusterCatalog, err error) error {
+	var unrecov *catalogderrors.Unrecoverable
+	if err == nil {
+		catalog.Status.ResolvedSource = nil
```
Why are we setting this to nil here?

anik120 replied:
Oops, probably a hangover from one of the old functions I was modifying. Looks like Ankita fixed this.

m1kola commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -168,7 +163,8 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
 		}
 
-		updateStatusUnpacked(&catalog.Status, unpackResult, contentURL, catalog.Generation, lastUnpacked)
+		_ = updateStatusProgressing(catalog, nil)
```
Are we intentionally not checking error here? If yes, then we can just do that, right?
```suggestion
		updateStatusProgressing(catalog, nil)
```

ankitathomas replied:
updateStatusProgressing was returning the error that was passed in unchanged, modified the function to avoid this

everettraven commented on hack/scripts/demo-script.sh:
```diff
@@ -2,32 +2,24 @@
-kubectl apply -f https://github.com/operator-framework/catalogd/releases/latest/download/catalogd.yaml
-kubectl wait --for=condition=Available -n olmv1-system deploy/catalogd-controller-manager --timeout=60s
-sleep 10
+make run
```
I think the intention behind this demo is that it shows you the entire cycle from installing catalogd on your cluster to creating a `ClusterCatalog` resource and exploring the catalog contents. I think using `make run` effectively removes the "how to install catalogd" part of the demo.

Not a big deal since we are talking about consolidating operator-controller and catalogd into a single monorepo, but thought I would call out the original intention behind why this was doing what it was doing before this change

ankitathomas replied:
thanks for the context, I've reverted to the older version of the script and added another step to install cert-manager

everettraven reviewed (dismissed):
These changes LGTM, but we should probably get @grokspawn and @joelanford to do a final pass as well prior to merging

ankitathomas commented:
/retest

ankitathomas commented:
/retest

joelanford commented on README.md:
```diff
@@ -87,9 +87,14 @@ Procedure steps marked with an asterisk (`*`) are likely to change with future A
         Message:               
-        Reason:                Unpacking
+        Reason:                Progressing
         Status:                False
-        Type:                  Unpacked
+        Type:                  Succeeded
```
This should be `Reason: Succeeded`, `Type: Progressing`, right?

joelanford commented on README.md:
```diff
@@ -87,9 +87,14 @@ Procedure steps marked with an asterisk (`*`) are likely to change with future A
         Status:                False
-        Type:                  Unpacked
+        Type:                  Succeeded
+        Last Transition Time:  2023-06-23T18:35:13Z
+        Message:               
+        Reason:                Serving
+        Status:                True
+        Type:                  Available
```
Just double checking that this is how the `kubectl describe` output looks like when there are two conditions? (It is hard to tell where one condition stops and the next starts)

perdasilva replied:
Yeah - I hate that. I try to anchor myself on the last transition time XD

anik120 replied:
Looks like Ankita did a whole copy paste of an actual output in this section, so that should reflect accurately what actually shows up. Resolving this for now, please reopen if you folks still have concerns.

joelanford commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -180,73 +180,60 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
+			Type:    v1alpha1.TypeProgressing,
+			Status:  metav1.ConditionFalse,
+			Reason:  v1alpha1.ReasonSucceeded,
+			Message: fmt.Sprintf("Successfully unpacked and stored content from %s", catalog.Spec.Source.Image.Ref),
```
Do we already reference the served image in the "Serving" condition? If so, I think I would suggest a simpler message like "Serving desired content" or (if polling is active) "Waiting until next poll interval"

@everettraven and @grokspawn  thoughts?

ankitathomas replied:
```
meta.SetStatusCondition(&status.Conditions, metav1.Condition{
	Type:    v1alpha1.TypeServing,
	Status:  metav1.ConditionTrue,
	Reason:  v1alpha1.ReasonAvailable,
	Message: fmt.Sprintf("Content from %s is being served", status.ResolvedSource.Image.Ref),
})
```
```
meta.SetStatusCondition(&catalog.Status.Conditions, metav1.Condition{
	Type:    v1alpha1.TypeProgressing,
	Status:  metav1.ConditionFalse,
	Reason:  v1alpha1.ReasonSucceeded,
	Message: fmt.Sprintf("Successfully unpacked and stored content from %s", catalog.Spec.Source.Image.Ref),
})
```
Looks like we're using the resolved image ref on one and the image ref from the spec on the other. 

We may want to have the resolved image ref on just the `Serving` condition like @joelanford suggested

perdasilva replied:
Do we need the image ref in any condition message at all? Would the image resolution information not be sufficient to derive that?

everettraven replied:
I think the additional information in the status fields outside of the conditions should be sufficient to convey what is being served and what was resolved

joelanford commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -180,73 +180,60 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
-		Message: result.Message,
-	})
-}
+func updateStatusProgressing(catalog *v1alpha1.ClusterCatalog, err error) {
```
```suggestion
func updateStatusProgressing(*v1alpha1.ClusterCatalog, error) {
```

Use explicit returns: https://go.dev/tour/basics/7

perdasilva replied:
would it be worth refactoring the package name to `catd` or `catdv1alpha1`?

everettraven replied:
Just a note that these look to be input parameters to the function, not return values

everettraven replied:
Ah I see why @joelanford may have thought that these were named return values. We are doing "naked returns" in this function. To reduce potential for confusion, maybe we should do something like:
```go
func updateStatusProgression(catalog *v1alpha1.ClusterCatalog, err error) {
        progressingCond := metav1.Condition{
                Type:  v1alpha1.TypeProgressing,
                Status: metav1.ConditionFalse,
                Reason: v1alpha1.ReasonSucceeded,
                Message: "...",
       }
       
       if err != nil {
               progressingCond.Status = metav1.ConditionTrue
               progressingCond.Reason = v1alpha1.ReasonRetrying
               progressingCond.Message = err.Error()
       }
       
       if errors.As(err, &unrecov) {
               progressingCond.Status = metav1.ConditionFalse
               progressingCond.Reason = v1alpha1.ReasonUnrecoverable
       }
       
       meta.SetStatusCondition(&catalog.Status.Conditions, progressingCond)
}
```

joelanford replied:
Oh yeah, @everettraven is right. I saw the naked returns and the error parameter and my mind jumped to named returns. My bad!

joelanford commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -180,73 +180,60 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
 
-func updateStatusStorageDeleteError(status *v1alpha1.ClusterCatalogStatus, err error) error {
+func updateStatusNotServing(status *v1alpha1.ClusterCatalogStatus) {
+	status.ContentURL = ""
```
Also unset `lastUnpacked` (and any other related fields)?

ankitathomas replied:
If we unpacked at any point in time, the only time we currently set `Serving` to false is during a delete, if we fail at removing unpacked contents. 

We can unset `lastUnpacked`, `ObservedGeneration` and `ResolvedSource` in the mean time, but we should be able to stop serving contents by just removing the `ContentURL`

everettraven replied:
Is there a potential use case where a user may want to still know the information conveyed by related fields if we are no longer serving? If so, I think I would keep them set. If we can't think of any I would agree with @joelanford that we should go ahead and unset them

joelanford replied:
Do we think that users might think that `lastUnpacked` carries an implication that something is still unpacked _right now_? 

That's what I was getting at here. But I do see both sides of this.

Are we keeping `ObservedGeneration` in the root? I can't remember, but I thought that was being removed.

anik120 replied:
>Are we keeping ObservedGeneration in the root? I can't remember, but I thought that was being removed. 

Not in this PR, that's unrelated. I think I remember agreeing we'd like to remove it, but we still wanted to confirm that removal of this field is acceptable/wanted. 
Any which way, that's a separate PR. 


>Is there a reasonable use case where a user may want to still know the information conveyed by related fields if we are no longer serving? If so, I think I would keep them set.

> Do we think that users might think that lastUnpacked carries an implication that something is still unpacked right now?

Only argument I can think of, is that "has the content been Unpacked?" is a separate question than "is the content being served". So the recipe`updateStatusNotServing` should only be concerned with answering "no the content is not being served any more", and nothing else. 
However, that's pedagogical :)

joelanford replied:
If we still had a separate `Unpacked` condition, I might agree. But now that we're down to `Serving` and `Progressing`, that's not a question we are answering in our API.

That makes me question the `lastUnpacked` field, actually. If conceptually, the field conveyed "the last time the served content changed", then maybe it follows that we would unset it when we stopped serving?

anik120 replied:
Resolving this since `lastUnpacked` is being unset now.

joelanford commented on internal/controllers/core/clustercatalog_controller_test.go:
```diff
@@ -250,19 +159,19 @@ func TestCatalogdControllerReconcile(t *testing.T) {
 		{
-			name:      "valid source type, unpack returns error, status updated to reflect error state and error is returned",
+			name:      "valid source type, unpack returns unrecoverable error, status updated to reflect unrecoverable error state and error is returned",
 			shouldErr: true,
```
Shouldn't this be `shouldErr: false`?

If the error is unrecoverable, the intention is to _not_ enter backoff retries.

everettraven replied:
Unless we've changed the logic, I think that we still expect `reconcile()` to return an error which is what we are unit testing here. Maybe the test function name is a bit misleading?

joelanford replied:
Oh.. hmm. Any reason we aren't testing `Reconcile` so that we can pull in the extra bit of logic it does in our unit tests?

Alternately, do we have another set of unit tests for `Reconcile`? That could probably work too.

everettraven replied:
testing `Reconcile` requires mocking kubernetes clients. where possible, I think we should avoid that due to the inherent complexity of mocking Kube

joelanford commented on internal/controllers/core/clustercatalog_controller_test.go:
```diff
@@ -117,107 +114,19 @@ func TestCatalogdControllerReconcile(t *testing.T) {
-							Reason: catalogdv1alpha1.ReasonUnpacking,
+							Type:   catalogdv1alpha1.TypeProgressing,
+							Status: metav1.ConditionTrue,
+							Reason: catalogdv1alpha1.ReasonRetrying,
```
This case should be unrecoverable.

joelanford commented on internal/controllers/core/clustercatalog_controller_test.go:
```diff
@@ -451,6 +375,16 @@ func TestCatalogdControllerReconcile(t *testing.T) {
 					},
 				},
+				Status: catalogdv1alpha1.ClusterCatalogStatus{
+					ContentURL: "URL",
```
Include any timestamps that would have been set when we polled and unpacked in the past?

We would expect those to be unset as well.

perdasilva commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -180,73 +180,60 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
 		return ctrl.Result{RequeueAfter: requeueAfter}, nil
 	default:
-		return ctrl.Result{}, updateStatusUnpackFailing(&catalog.Status, fmt.Errorf("unknown unpack state %q: %v", unpackResult.State, err))
+		panic(fmt.Sprintf("unknown unpack state %q", unpackResult.State))
```
I'm guessing this will kill the pod. Is it by design to stop reconciliation for all CEs if one gets into this state?

everettraven replied:
The thought process behind the panic here is that this is a programmer error as the only state that can be returned from the existing `Unpacker` implementation is `source.StateUnpacked`. There is nothing a user can do to resolve this and since this is a programmer error I think panicking is appropriate.

anik120 replied:
@perdasilva resolving this, please reopen if your question wasn't adequately answered.

everettraven commented on internal/controllers/core/clustercatalog_controller_test.go:
```diff
@@ -507,9 +451,9 @@ func TestCatalogdControllerReconcile(t *testing.T) {
-							Type:   catalogdv1alpha1.TypeDelete,
-							Status: metav1.ConditionFalse,
-							Reason: catalogdv1alpha1.ReasonStorageDeleteFailed,
+							Type:   catalogdv1alpha1.TypeProgressing,
```
Yeah, that is what I would expect based on the behaviors we implemented in the ClusterCatalogController

joelanford replied:
Let's put that in the test then.

anik120 replied:
Looks like this was added in the latest push Ankita did, resolving this conversation manually, since "Done"

anik120 commented:
@ankitathomas thoughts on squashing the commits? (we'll be coauthors of one giant commit 😃)

anik120 commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -174,73 +174,54 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
+func updateStatusNotServing(status *v1alpha1.ClusterCatalogStatus) {
 	status.ResolvedSource = nil
+	status.ContentURL = ""
+	status.ObservedGeneration = 0
```
Should we be modifying this here? Is `0` the equivalent of `unset`?

ankitathomas replied:
0 is the default value for int64, so that should be what an unset value is since ObservedGeneration isn't a pointer

anik120 committed:
⚠ Status condition clean up

[RFC: ClusterCatalog Status Conditions](https://docs.google.com/document/d/1Kg8ovX8d25OqGa5utwcKbX3nN3obBl2KIYJHYz6IlKU/edit) implementation.

Closes [363](operator-framework#363)
Closes [364](operator-framework#364)
Closes [365](operator-framework#365)

Co-authored-by: Ankita Thomas <ankithom@redhat.com>
Co-authored-by: Anik Bhattacharjee <anbhatta@redhat.com>

anik120 commented:
@ankitathomas fyi I removed the changes to the demo scripts and related stuff. The cognitive load required for review was probably getting a little heavier because of those changes, at least for me. We'll most definitely have more high quality reviews on the changes you want to make there if we put it up as a separate PR anyway.

anik120 commented:
@ankitathomas @joelanford @everettraven ready for review again.

everettraven commented on docs/fetching-catalog-contents.md:
```diff
@@ -84,12 +84,8 @@ of a catalog can be read from:
-      reason: UnpackSuccessful
-      status: "True"
-      type: Unpacked
+    .
+    .
```
(Optional) Nit: Should this still show the conditions?

anik120 replied:
+1 Todd. Opened https://github.com/operator-framework/catalogd/issues/410

tmshort replied:
I had already fixed the issues with the service name in #406, but the self signed fix still needs to be done.

everettraven commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -118,10 +118,13 @@ func (r *ClusterCatalogReconciler) reconcile(ctx context.Context, catalog *v1alp
+			updateStatusProgressing(catalog, err)
+			return ctrl.Result{}, err
 		}
+		updateStatusNotServing(&catalog.Status)
```
Could we add a comment here that talks about the issue that we encountered when we both modified the status and the `ClusterCatalog` finalizers? I think that issue still exists, we just won't hot-loop because of Joe's change and we would end up setting the status to the same thing (i.e no change) so we end up successfully removing the finalizer on the second reconcile.

anik120 replied:
Created https://github.com/operator-framework/catalogd/issues/409. I don't think we need a TODO comment here besides the issue

joelanford commented on test/tools/imageregistry/pre-upgrade-setup.sh:
```diff
@@ -31,4 +31,4 @@ spec:
 EOF
 
-kubectl wait --for=condition=Unpacked --timeout=60s ClusterCatalog $TEST_CLUSTER_CATALOG_NAME
+kubectl wait --for=condition=Serving=true --timeout=60s ClusterCatalog $TEST_CLUSTER_CATALOG_NAME
```
I think this needs to be reverted (for now) because the most recent release is still using the `Unpacked` condition type and this code runs for the most recent release.

Once this code merges AND we cut a release, we'll need to follow back up with this change.

We should do it this way so that we stay in the habit of having green upgrade-e2e runs and when the do turn red, it is because of something _in the particular PR_, not something that merged (or didn't merge in a prior PR).

anik120 replied:
I had the same thought too [here](https://github.com/operator-framework/catalogd/pull/379#issuecomment-2334474770). I think @ankitathomas had a clearer picture with what to do with this change, but tbh making this change here is confusing for me too. @ankitathomas reverting this so that we can handle it in a follow up PR.

anik120 commented:
`codecov/project` is complaining about a test not being included [for the panic being introduced](https://app.codecov.io/gh/operator-framework/catalogd/pull/379?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=operator-framework#da31cb2031f8df66e18f98829deddd10-R176)

tmshort commented on README.md:
```diff
@@ -40,58 +40,48 @@ Procedure steps marked with an asterisk (`*`) are likely to change with future A
+        Reason:                Available
+        Status:                True
+        Type:                  Serving
+      Content URL:             https://catalogd-catalogserver.olmv1-system.svc/catalogs/operatorhubio/all.json
```
I'm gonna whine about this (new service name), but it can be fixed in #410.

tmshort replied:
I ended up fixing it.

tmshort committed:
Handle new finalizer code

Signed-off-by: Todd Short <todd.short@me.com>

joelanford commented on internal/controllers/core/clustercatalog_controller.go:
```diff
@@ -308,10 +280,13 @@ func NewFinalizers(localStorage storage.Instance, unpacker source.Unpacker) (crf
 		if err := localStorage.Delete(catalog.Name); err != nil {
-			return crfinalizer.Result{}, updateStatusStorageDeleteError(&catalog.Status, err)
+			updateStatusProgressing(catalog, err)
+			return crfinalizer.Result{}, err
```
I think we need to return `crfinalizer.Result{StatusUpdated: true}` on any code path that changes the status conditions. Otherwise, I'm not sure we fall into the `if finalizeResult.Updated || finalizeResult.StatusUpdated` block in line 130.

tmshort committed:
Set StatusUpdated to true for finalizers

Signed-off-by: Todd Short <todd.short@me.com>
